Improve allocations in WinForms startup

JeremyKuhne opened the pull request:
In a blank WinForms app a large percentage of the allocations on startup are coming from System.Drawing and Microsoft.Win32.SystemEvents (25%+).

One of the biggest culprits is the delegate system in System.Drawing that was put in place to handle multiple names for libgdiplus. As Windows doesn't have an issue with multiple dll names I've moved this exclusively to the Unix code. One other option for Unix (that would allow sharing the definitions again) would be to include the same DllImports twice under different namespaces with a different definition for the library name. (Well over 10% of the allocations are in this space.)

Another problem is the HandleCollector class. This was put in place to force GC collections to attempt to get forgotten GDI handles to avoid running out of them. This *was* a big problem in Windows 98, and much less so in NT4 and Windows 2000. GDI handles now come from one pool of 65 thousand handles per process (as of Windows XP). The 9x limits are long gone- this has become nothing more than overhead.

WindowClass creation was also creating unnecessary overhead (in SystemEvents). It was doing some unnecessary contortions to allow for multiple appdomains, which are no longer an issue. Native signatures were also suboptimal.

Lastly there was a fair amount of overhead on startup through Icon (getting the default icon from the app resource stream) and some through non-blittable fonts. Exposure of LOGFONT through object was also subject to a number of holes (size and null were never checked).

There are similar changes I've made in WinForms (notably with WindowClass creation). Together with those startup allocations drop from about 5,000 to around 1,500. Total bytes allocated drop by about 20%. Speed is really tough to measure as there is quite a bit of variance due to the fact so much is happening through Windows message pumping.

I still need to finish the Unix changes, just a little bit of busy work. Getting this up for discussion/review of the approach and impact. I'll link the WinForms PR once I have it up.

qmfrederik commented:
Interestingly, this PR is doing the exact opposite of #38996 - diverging the Windows & Unix implementations instead of converging them.

On Unix, the difficulty is the the different names for the same library across distributions (e.g. `libgdiplus.so` vs `libgdiplus.so.0`). If that can be consolidated and we can move back to simple `[DllImport]` methods, then awesome.

Also - it looks like this PR is doing a couple of things at the same time - reducing allocations in System.Drawing on startup on Windows, general cleanup, additional validation,... .
I think it would be easier to reason about these changes if they were part of separate commits/PRs.

> Speed is really tough to measure as there is quite a bit of variance due to the fact so much is happening through Windows message pumping.

Well, you could for example make `LoadFunctionPointers` public and call it from a benchmark, that should give you a rough estimate about how much time is spent loading each pointer, right?

If this significantly impacts startup performance, would lazy-loading some or all function pointers be an option?

jkotas commented:
> As Windows doesn't have an issue with multiple dll names I've moved this exclusively to the Unix code.

Unix flavor built for netcoreapp does not have the issue either. It is only the legacy netcoreapp2.0 / netstandard version the has this problem. It would be best to keep the delegates specific to Unix netcoreapp2.0 / netstandard versions. I expect they are going to be deleted once we stop caring about these legacy targets.

JeremyKuhne commented:
> it can use the new netcoreapp3.0 APIs that resolve the DllImport targets dynamically

I missed this feature. :) @jkoritzinsky do you have links on this?

>  It is only the legacy netcoreapp2.0 / netstandard version the has this problem

I wonder if we can just pull the built version for this target? That would make this a very nice/clean change @ericstj/@safern?

> I think it would be easier to reason about these changes if they were part of separate commits/PRs.

I intend to break it out (at the very least the SystemEvents). We're running short on time for 3.0 so I wanted to get something up to have discussions around while I iterate further.

> that should give you a rough estimate about how much time is spent loading each pointer

That I can do, and this would definitely be faster, but probably not huge in the overall scheme of things. For my comment I was referring to impact to the total time to show the form. The total time I'm seeing showing a form is a bit over 100ms on an NVMe drive, but has a huge delta from run to run. The bigger issue here is the number of allocations, which also keep coming with the handle collector. There is the ongoing tax of making calls through both of these mechanisms as well. The handle collector makes GCs more likely on two fronts, (1) additional allocations and (2) unnecessary forced GCs.

Here is the app I was using to look at things.

``` C#
    static class Program
    {
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);

            Form form = new Form
            {
                Text = "My Very Own Form",
                BackColor = Color.CornflowerBlue
            };

            Application.Run(form);
        }
    }
```

When measuring "wall time":

``` C#
    static class Program
    {
        static Stopwatch s_stopwatch = new Stopwatch();
        static long s_totalBytes;
        static long s_threadBytes;

        [STAThread]
        static void Main()
        {
            s_totalBytes = GC.GetTotalAllocatedBytes(precise: true);
            s_threadBytes = GC.GetAllocatedBytesForCurrentThread();
            s_stopwatch.Start();

            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);

            Form form = new Form
            {
                Text = "My Very Own Form",
                BackColor = Color.CornflowerBlue
            };

            form.Shown += Form_Shown;

            Application.Run(form);
        }

        private static void Form_Shown(object sender, EventArgs e)
        {
            s_stopwatch.Stop();
            long threadBytes = GC.GetAllocatedBytesForCurrentThread();
            long currentBytes = GC.GetTotalAllocatedBytes(precise: true);
            MessageBox.Show($"{s_stopwatch.ElapsedMilliseconds} ms, {threadBytes - s_threadBytes} thread bytes, {currentBytes - s_totalBytes} total bytes");
        }
    }
```

danmosemsft commented:
There's a lot in this change and it's several distinct changes. Would it make sense to keep it as separate commits (could still be in the same PR)? That way if issues emerge it's easier to bisect and possibly snip just one commit as needed - especially given as you say it's late. 

Not sure how painful it would be to go back and break into commits.

jkoritzinsky commented:
@JeremyKuhne you're looking for the `AssemblyLoadContext.ResolvingUnmanagedDll` event. That event takes in an assembly and a name provided via a `[DllImport]` attribute and expects a handle to the native library to be returned. You can use the methods on the `System.Runtime.InteropServices.NativeLibrary` class to load the native library and get the handle.

jkoritzinsky commented:
You could also call [`System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary.setdllimportresolver?view=netcore-3.0#System_Runtime_InteropServices_NativeLibrary_SetDllImportResolver_System_Reflection_Assembly_System_Runtime_InteropServices_DllImportResolver_) which would likely better fit your scenario by encapsulating some of the logic.

JeremyKuhne commented:
> Would it make sense to keep it as separate commits

I will rebuild it into separate commits where it isn't extremely painful. :)

jkotas commented:
`NativeLibrary.SetDllImportResolver` is the right API to use here. A good place to call it is explicit static constructor of the NativeMethods type, so that the callback is registered before any of the DllImports is resolved.

Tanya-Solyanik commented on src/Common/src/Interop/Windows/User32/Interop.CreateWindowEx.cs:
```diff
@@ -9,8 +9,19 @@ internal partial class Interop
-        [DllImport(Libraries.User32, CharSet=CharSet.Unicode, SetLastError=true, BestFitMapping=true, ExactSpelling =true)]
-        public static extern IntPtr CreateWindowExW(int exStyle, string lpszClassName, string lpszWindowName, int style, int x, int y, int width,
-                                              int height, IntPtr hWndParent, IntPtr hMenu, IntPtr hInst, IntPtr pvParam);
+        [DllImport(Libraries.User32, CharSet=CharSet.Unicode, SetLastError=true, ExactSpelling =true)]
```
nit: extra space in `ExactSpalling =true`

JeremyKuhne commented:
Moved the `SystemEvents` changes to #39312. Changed this to use the new resolver callback on Unix. I haven't tested the Unix stuff yet- @jkoritzinsky does it look like the right setup?

Now it's easier to see the diff with the APIs.

jkoritzinsky commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -90,1016 +92,612 @@ private static IntPtr LoadNativeLibrary()
-                LoadFunctionPointers();
+                NativeLibrary.SetDllImportResolver(Assembly.GetExecutingAssembly(), DllImportResolver);
+
+                s_gdipModule = LoadNativeLibrary();
```
Is `s_gdipModule` still used anywhere? The only usage I see of it is here.

JeremyKuhne replied:
I'll double check

jkoritzinsky commented on src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:
```diff
@@ -627,13 +533,8 @@ public static IntPtr CreateDIBSection(HandleRef hdc, ref NativeMethods.BITMAPINF
-        {
-            return System.Internal.HandleCollector.Add(IntCreateRectRgn(x1, y1, x2, y2), SafeNativeMethods.CommonHandles.GDI);
-        }
+        [DllImport(ExternDll.Gdi32, SetLastError = true, ExactSpelling = true)]
```
Does the GDI32 library have the same issues as GDI+? If it doesn't, then this should be fine.

JeremyKuhne replied:
It doesn't

jkoritzinsky commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -90,1016 +92,612 @@ private static IntPtr LoadNativeLibrary()
+                    return LoadNativeLibrary();
+                }
+
+                return NativeLibrary.Load(libraryName, assembly, searchPath);
```
You should `IntPtr.Zero` to continue the regular library load logic. This change would cause any native library loaded by this assembly to resolve the native library next via the `DllImportSearchPath`s instead of via the `AssemblyLoadContext`'s `ResolvingUnmanagedDll` method if this assembly is loaded into a custom `AssemblyLoadContext`.

JeremyKuhne replied:
Good to know, thanks! I'll update it.

jkoritzinsky commented on src/System.Drawing.Common/src/System/Drawing/Printing/LibcupsNative.cs:
```diff
@@ -9,10 +9,10 @@ namespace System.Drawing.Printing
 {
     internal static class LibcupsNative
```
You should add a static constructor here that calls `PlatformInitialize` to set up the `DllImportResolver`. Without that, it's possible for something to call a P/Invoke here without that set up yet.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/LibraryResolver.cs:
```diff
@@ -0,0 +1,32 @@
+
+        private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
+        {
+            if (LibraryLoaders.TryGetValue(libraryName, out Func<IntPtr> loader))
```
I do not see anything that guarantees thread safety here. I do not think this needs to be a fancy registration scheme. Just check for the two libraries we care about and call the appropriate method

JeremyKuhne replied:
How do you suggest I ensure we are hooked dynamically when the relevant wrapping classes are accessed? I was relying on the dictionary access from the static constructors from the two classes. I'm also unclear on how threading would bust as it currently is set up.

JeremyKuhne replied:
My presumption is that the callback results are cached for positive results. Is that true @jkoritzinsky?

jkotas replied:
> I'm also unclear on how threading would bust as it currently is set up.

The two static constructors that register the LibraryLoaders can run in parallel on different threads; or they can run in parallel while the other thread is resolving the library.

```
class LibraryResolver
{
     static LibraryResolver()
     {
         NativeLibrary.SetDllImportResolver(typeof(LibraryResolver).Assembly, DllImportResolver);
     }

      private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
      {
           if (LibcupsNative.LibraryName == "libcups") return LibcupsNative.LoadLibcups();
           if (libraryName == "...") return ...
           return IntPtr.Zero;
      }

     void EnsureRegistered()
     {
          // dummy call to trigger the static constructor
     }
}
```

Then call `LibraryResolver.EnsureRegistered();` in the static constructor of each of `*Native` types that depend on it.

JeremyKuhne commented:
As to direct startup impact here: Making an initial call to `Graphics.FromHwnd(IntPtr.Zero);` we go from ~3600 allocations to 196. Time is cut from 32 to 12ms.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/LibraryResolver.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace System.Drawing
+{
+    class LibraryResolver
```
Nit: This can be `internal static` as before.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Printing/LibcupsNative.cs:
```diff
@@ -9,10 +9,14 @@ namespace System.Drawing.Printing
-        private static IntPtr LoadLibcups()
+        static LibcupsNative()
+        {
+            LibraryResolver.EnsureRegistered();
```
Do we need to call this also in `SafeNativeMethods.Gdip` and `UnsafeNativeMethods.Gdip` ?

JeremyKuhne replied:
Yeah, I feel particularly stupid now, I was trying to build the Linux bits and mistyped the property name so I was getting the Windows build and didn't notice. Derp.

JeremyKuhne commented:
Looking at NetFX WinForms we've allocated 5.6K objects in 406ms during startup, with the largest chunk (2.4K) coming from `System.Configuration`.  On Core we're at 4.9K in 414ms, with the vast majority coming from System.Drawing initializing it's import delegates.

My changes put this at 1.6K in 394ms. The vast majority of allocations removed are from this change. A few hundred are from SystemEvents and removing HandleCollector / optimizing Window registration in WinForms.

JeremyKuhne commented:
@safern, can you take a look at this? I think it is in good shape now.

safern commented on src/System.Drawing.Common/System.Drawing.Common.sln:
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27213.1
+# Visual Studio Version 16
```
Was this intentional? We still support VS2017

JeremyKuhne replied:
VS does this. Doesn't block using downlevel, it's just a track of what last edited it.

safern commented on src/System.Drawing.Common/ref/System.Drawing.Common.csproj:
```diff
@@ -1,16 +1,16 @@
   <ItemGroup>
     <Compile Include="System.Drawing.Common.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsNetFx)' == 'true' OR '$(TargetsNETStandard)' == 'true'">
```
We no longer need this.

safern replied:
Let's cleanup the csproj for netfx, netcoreap2.0 and netstandard2.0 stuff.

JeremyKuhne replied:
Sure, note that I'm adding back uap as some of the other test projects depend on it.

safern commented on src/System.Drawing.Common/ref/System.Drawing.Common.csproj:
```diff
@@ -1,16 +1,16 @@
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'true' OR '$(TargetsUap)' == 'true'">  
+  <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'true' OR '$(TargetsUap)' == 'true'">
     <Compile Include="System.Drawing.Common.Forwards.cs" />
```
We can remove this and Uncondition `IsPartialFacadeAssembly`

safern replied:
We no longer need this manual forwards if you set `IsPartialFacadeAssembly` to true.

JeremyKuhne replied:
I'm confused. `IsPartialFacadeAssembly` was only on for `NetFx`. Are you saying we should always have it on now? Why are we able to do that, if so?

safern replied:
I think it was because of the way the ref assembly was composed, but now that we are building it live only for 3.0 and uap, we can now use IsPartialFacadeAssembly, but no need to fix it. Either way works fine.

safern commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,55 @@ public override string ToString()
+                int fullLength = logFont.lfFaceName.Length;
+                Span<char> temp = stackalloc char[fullLength];
+                temp[0] = '@';
+                logFont.lfFaceName.Slice(0, fullLength - 1).CopyTo(temp.Slice(1));
```
This will drop the last character of the font name, is that expected?

JeremyKuhne replied:
Yeah, you have to as the buffer is a fixed size. Sometimes that may cause a final character to drop (there may be nulls if the name is shorter than the max).

safern replied:
Ah I see, makes sense.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,55 @@ public override string ToString()
+                Span<char> temp = stackalloc char[fullLength];
+                temp[0] = '@';
+                logFont.lfFaceName.Slice(0, fullLength - 1).CopyTo(temp.Slice(1));
+                temp.CopyTo(logFont.lfFaceName);
```
Span copy operations do handle overlapping buffers. You should not need the extra buffer on stack to do this.

JeremyKuhne replied:
Cool, did not know that. I'll update the code.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,55 @@ public override string ToString()
+                int fullLength = logFont.lfFaceName.Length;
+                Span<char> temp = stackalloc char[fullLength];
+                temp[0] = '@';
+                logFont.lfFaceName.Slice(0, fullLength - 1).CopyTo(temp.Slice(1));
```
If the original `lfFaceName` was `fullLength - 1` + null terminator, the modified `lfFaceName` is going to be missing the null terminator. Is it a problem? The docs say that `lfFaceName` should be null terminated.

JeremyKuhne replied:
The buffer is fixed, so I don't think so? We certainly didn't check before. I'll double check.

JeremyKuhne replied:
I don't think it really needs it, but the docs don't give the wiggle room. Checking my machine with > 1000 font results I don't have any that totally fill the buffer (i.e. all are null terminated). I'll force the null in. I suspect that this code has always been broken, I don't see any reference to the APIs allowing font names to be preceded with `@`. I suspect what we _should_ be doing is rolling the `@` back out of the returned `LOGFONT` and passing it back in. I'll create an issue to investigate this.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,57 @@ public override string ToString()
+            // Prefix the string with '@' if this is a gdiVerticalFont.
+            if (_gdiVerticalFont)
+            {
+                int fullLength = logFont.lfFaceName.Length;
```
Nit: Make this `int fullLength = faceName.Length;` and move this one line below to avoid constructing the Span twice? Or just use `faceName.Length` instead of the local?

safern commented on src/System.Drawing.Common/src/System/Drawing/Icon.Windows.cs:
```diff
@@ -159,42 +160,36 @@ void ISerializable.GetObjectData(SerializationInfo si, StreamingContext context)
             if (!File.Exists(filePath))
             {
-                throw new FileNotFoundException(filePath);
+                throw new FileNotFoundException(new FileNotFoundException().Message, filePath);
```
Why is using the default's constructor message better than the one that gets constructed for when the message is null?

safern commented on src/System.Drawing.Common/src/System/Drawing/Icon.Windows.cs:
```diff
@@ -216,26 +211,32 @@ public IntPtr Handle
+                            new HandleRef(null, info.hbmMask),
+                            sizeof(SafeNativeMethods.BITMAP),
+                            ref bitmap);
+                        _iconSize = new Size((int)bitmap.bmWidth, (int)(bitmap.bmHeight / 2));
```
Why are the casts necessary now?

JeremyKuhne replied:
Because I changed the struct to the native uint. Felt it was better to express that there is no "negative" in the native representations.

safern replied:
Ok cool.

safern commented on src/System.Drawing.Common/src/System/Drawing/LibraryResolver.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        static LibraryResolver()
+        {
+            // Hook our custom resolver
```
Nit: is this comment really necessary?

JeremyKuhne replied:
I guess not

safern commented on src/System.Drawing.Common/src/System/Drawing/Printing/LibcupsNative.cs:
```diff
@@ -9,10 +9,14 @@ namespace System.Drawing.Printing
+        internal static IntPtr LoadLibcups()
         {
             // We allow both "libcups.so" and "libcups.so.2" to be loaded.
 #if netcoreapp20
```
We can now remove all this if defs.

safern commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -425,9 +435,17 @@ private void ExtractAssociatedIcon_FilePath_Success(string filePath)
 
         [ActiveIssue(20884, TestPlatforms.AnyUnix)]
         [ConditionalFact(Helpers.IsDrawingSupported)]
```
Does this still Apply?

safern commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -425,9 +435,17 @@ private void ExtractAssociatedIcon_FilePath_Success(string filePath)
 
         [ActiveIssue(20884, TestPlatforms.AnyUnix)]
```
Can we do something similar in Icon.Unix.cs when the file doesn't exist?

JeremyKuhne replied:
Probably. The Icon class can be further unified I think. As we fix the rest of the P/Invokes it will be much easier to unify the implementations.

safern replied:
Cool. We can follow up after this PR on that then.

JeremyKuhne commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,70 @@ public override string ToString()
+
+            SafeNativeMethods.LOGFONT nativeLogFont = ToLogFontInternal(graphics);
+
+            if (!type.IsValueType)
```
@jkotas I changed this so round tripping blittable value types works. Is there a better way to do this?

JeremyKuhne replied:
So either type below works now:

``` C#
[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
public unsafe struct LOGFONT
{
    private const int LF_FACESIZE = 32;

    public int lfHeight;
    public int lfWidth;
    public int lfEscapement;
    public int lfOrientation;
    public int lfWeight;
    public byte lfItalic;
    public byte lfUnderline;
    public byte lfStrikeOut;
    public byte lfCharSet;
    public byte lfOutPrecision;
    public byte lfClipPrecision;
    public byte lfQuality;
    public byte lfPitchAndFamily;
    private fixed char _lfFaceName[LF_FACESIZE];
}

[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
public unsafe class LOGFONTCLASS
{
    private const int LF_FACESIZE = 32;

    public int lfHeight;
    public int lfWidth;
    public int lfEscapement;
    public int lfOrientation;
    public int lfWeight;
    public byte lfItalic;
    public byte lfUnderline;
    public byte lfStrikeOut;
    public byte lfCharSet;
    public byte lfOutPrecision;
    public byte lfClipPrecision;
    public byte lfQuality;
    public byte lfPitchAndFamily;
    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 32)]
    public string lfFaceName;
}
```

jkotas replied:
The only way to maintain full compatibility with what used to be there is to keep the `AsAny` marshaling. I think anything else it going to be broken for some set of cases that used to work before.

jkotas replied:
> So either type below works now:

Does this work?

```
[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
public struct LOGFONT
{
    private const int LF_FACESIZE = 32;

    public int lfHeight;
    public int lfWidth;
    public int lfEscapement;
    public int lfOrientation;
    public int lfWeight;
    public byte lfItalic;
    public byte lfUnderline;
    public byte lfStrikeOut;
    public byte lfCharSet;
    public byte lfOutPrecision;
    public byte lfClipPrecision;
    public byte lfQuality;
    public byte lfPitchAndFamily;
    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 32)]
    public string lfFaceName;
}
```

jkotas replied:
If it is important for AsAny marshaling on bllitable types to be faster, the runtime is best place to fix it. I think you trying to special case here is losing battle unless you are willing to resort to breaking changes.

Having said this, I doubt that the AsAny marshaling overhead would show up here.

JeremyKuhne replied:
Ok, I'm going for breaking here (unblittable value types won't work). I finished changing the `LOGFONT` references and there is just too much benefit to keeping it blittable internally. We can add a P/Invoke overload in the future if we really need to but I'd rather not put that in unless we really need it.

Finishing tidying up the `LOGFONT` references I see a significant improvement in getting `SystemFonts.DefaultFont` (which is in WinForms startup). Just getting that property goes from 33ms/3920bytes down to 22ms/2000bytes.

`SystemFonts.MenuFont` (which WinForms uses) goes from 38ms/4192bytes down to 24ms/1936bytes.

jkotas replied:
What is the impact of this breaking change? Performance improvement alone is not sufficient for doing a breaking change.

JeremyKuhne replied:
> What is the impact of this breaking change? 

If, for some reason, you created your own data structure for this that was a non-blittable struct you would no longer work. We never documented the internals of the object returned- if you grab and return it through the APIs then you will still work. If you grab an old versions and somehow get it to a new version, it will work- and vice-versa.

jkotas replied:
Agree that this looks like benign breaking. I assume that you are going to apply similar cleanup to the WinForms side as well - so that the AsAny marshalling is eliminated there too.

JeremyKuhne commented:
> Why do you need to specify the HintPath? This should not have a HintPath

I don't. The project system did it and I missed it. :)

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,70 @@ public override string ToString()
+            if (!type.IsValueType)
+            {
+                // Unfortunately PtrToStructure requires that the passed in object not be a value type.
+                // The underlying runtime code supports it, and perhaps there is another way to do this.
```
Blasting a new content over a boxed valuetype is unsafe operation. That's why the `PtrToStructure` APIs are designed they way they are designed - to make them somewhat more safe. I think this comment can be deleted. It is pretty unlikely we would ever try to fix that in these legacy APIs.

JeremyKuhne replied:
Good to know, I'll update it.

jkotas commented on src/System.Drawing.Common/src/System/Drawing/Font.cs:
```diff
@@ -275,8 +275,70 @@ public override string ToString()
+
+            if (!type.IsValueType)
+            {
+                // Unfortunately PtrToStructure requires that the passed in object not be a value type.
```
Nit: This comment may rather be in front of the IsValueType condition

jkotas commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -375,6 +350,7 @@
     <Reference Include="System.ObjectModel" />
     <Reference Include="System.Resources.ResourceManager" />
     <Reference Include="System.Runtime" />
+    <Reference Include="System.Runtime.CompilerServices.Unsafe" />
```
Is this used anywhere in the current version of the changes? I am not able to find any references to it.

JeremyKuhne replied:
No, good catch. :) Removed.

RussKie commented on src/System.Drawing.Common/src/System/Drawing/Text/PrivateFontCollection.cs:
```diff
@@ -34,7 +34,7 @@ protected override void Dispose(bool disposing)
                 try
                 {
 #if DEBUG
```
🤔 Do we need these #defs at all?
